### PR TITLE
refac: dulicate events does not affect amount of periods

### PR DESCRIPTION
**Pull request by lasrinnil**
## Description

We are guaranteed to get all events from the `ServiceBus` at least once, but theoretically, we could get it multiple times. That events are present multiple times should not affect the number of periods that are created for a given event type.

This PR will address this and add a test to verify it.

## References

* #32

### CODE REVIEW

File: source/databricks/package/balance_fixing_total_production.py [hunk 1]
```diff
@@ -189,7 +188,7 @@ def _get_metering_point_periods_df(
     )  # Only aggregate when metering points is connected
     metering_point_periods_df = metering_point_periods_df.where(
         col("MeteringPointType") == MeteringPointType.production.value
-    )
+    ).distinct()
```
Comment (BjarkeMeier) [inline source/databricks/package/balance_fixing_total_production.py:191]: I'm not sure this works in practice. Consider all the selected data of the dataframe below. Are you sure they will be identical when the same message is being received more than once?

image

Also, I don't think it's a good idea from a performance perspective to inspect all data in all columns. I recommend using `dropDuplicates()` on a specific subset of columns. At least if you want to do it here.

But if you end up deciding that the subset of columns to decide duplication upon matches the final select of this function then you can simply move your `distinct()` to after the final select. That is my recommendation.

@seblapela, do you have anything interesting to add about this topic?

File: source/databricks/tests/integration/calculator/balance_fixing_total_production/test_get_metering_points_df.py [hunk 2]
```diff
@@ -450,3 +450,31 @@ def test__metering_points_are_periodized_by_effective_date(
 
     # Assert
     assert (actual_df.count() == 1) == expected
+
+
+def test__duplicate_events_does_not_affect_amount_of_periods(
```
Comment (BjarkeMeier) [inline source/databricks/tests/integration/calculator/balance_fixing_total_production/test_get_metering_points_df.py:455]: I don't think you should only test that it doesn't affect _the amount of_ periods. The periods should still be equal. So I suggest the more precise name:
```suggestion
def test__duplicate_events_does_not_affect_periods(
```

File: source/databricks/tests/integration/calculator/balance_fixing_total_production/test_get_metering_points_df.py [hunk 3]
```diff
@@ -450,3 +450,31 @@ def test__metering_points_are_periodized_by_effective_date(
...
+    grid_area_df,
+):
+    # Arrange
+    created_events_df = metering_point_created_df_factory(operation_time=second_of_june)
+    connected_events_df = metering_point_connected_df_factory(
+        operation_time=second_of_june
+    )
+    # multiple duplicate 'connected' events are added to the dataframe
+    integration_events_df = (
+        created_events_df.union(connected_events_df)
```
Comment (BjarkeMeier) [inline source/databricks/tests/integration/calculator/balance_fixing_total_production/test_get_metering_points_df.py:467]: Should we also test with multiple created events?

File: source/databricks/tests/integration/calculator/balance_fixing_total_production/test_get_metering_points_df.py [hunk 4]
```diff
@@ -450,3 +450,31 @@ def test__metering_points_are_periodized_by_effective_date(
...
+        .union(connected_events_df)
+        .union(connected_events_df)
+        .union(connected_events_df)
```
Comment (BjarkeMeier) [inline source/databricks/tests/integration/calculator/balance_fixing_total_production/test_get_metering_points_df.py:470]: Can `offSet`, `partition` or any other service bus meta data column be distinct when receiving the same message more than once?

**Review by BjarkeMeier — COMMENTED** [review]
Should we test duplicate grid area updated events as-well?

**Comment by lasrinnil** [comment]
> Should we test duplicate grid area updated events as-well?

Yes in a upcomming PR I will add this

File: source/databricks/package/balance_fixing_total_production.py (hunk 1 above)
Comment (lasrinnil) [reply source/databricks/package/balance_fixing_total_production.py:191]: @Sondergaard is `OperationCorrelationId` "strong" enough to depend on it always unique?

**Review by seblapela — APPROVED** [review]

**Review by BjarkeMeier — APPROVED** [review]
LGTM

**Pull request merged**